### PR TITLE
FEAT[BB-281]: Display Author of works in WorkTable

### PR DESCRIPTION
**Pull request by the-good-boy**
### Problem

This PR is for the ticket BB-281(https://tickets.metabrainz.org/browse/BB-281)

### Solution

I have created a middleware which will load up the author for all the works in an Edition. This middleware has been used after the relationship-set of the current edition has been loaded. From these relationships, we will filter the relationshipType=10(Edition contains work), and then load the relationship-set of the works. For the relationshipSet of a particular work, we only care about the relationshipType=8(Author has written this work) to get the author of this work.
WorkTable

### Areas of Impact

modified:   src/client/components/pages/entities/work-table.js
modified:   src/server/helpers/middleware.ts
modified:   src/server/routes/entity/edition.js

### CODE REVIEW

**Comment by the-good-boy** [comment]
On second thought, perhaps it would be better to close this pr and create a new one, so that it would be easier to review. Rebasing the codebase makes this look so messy. @MonkeyDo ?

**Comment by MonkeyDo** [comment]
If that's not too much work, I agree it would probably be easier to recreate the changes in a new PR.
Thanks for going back to this PR, and for thinking of me when I review ! :)

**Comment by MonkeyDo** [comment]
Alternatively, if the changes are minimal and you can tell me exactly which files to look at, it's doable.

**Comment by the-good-boy** [comment]
@MonkeyDo I have made changes to only 3 files. 
1.  **src/server/helpers/middleware.ts** : Here I have made a middleware named 'loadWorkTableAuthors' (line 107)
2. **src/server/routes/entity/edition.js** : I have used this middleware (line 240)
3. **src/client/components/pages/entities/work-table.js** : Created a column for displaying the Author name.
Hope it helps!

**Review by MonkeyDo — COMMENTED** [review]
The mechanism to retrieve the Authors looks good, and displaying them on the page is perfect, nicely done !

I'd like to see the contents of loadWorkTableAuthors moved to a function in the ORM instead.
These sort of tools that allow us to do powerful-but-complex things with entities are better implemented there, where they can be reused in other contexts (for example the API, etc.), and to a certain degree helps with not having to reimplement functions like `getEntityWithAlias` all over again (we can reuse [getEntity](https://github.com/metabrainz/bookbrainz-data-js/blob/master/src/func/entity.ts#L196)).

I'd say the right place for it would be a new file `src/func/work.ts`

**Comment by MonkeyDo** [comment]
P.S: You can still have a `loadWorkTableAuthors` middleware if that's easier, but in turn that would call the function on the ORM.

File: src/client/components/pages/entities/work-table.js [hunk 1]
```diff
@@ -59,6 +59,7 @@ function WorkTableRow({showAddedAtColumn, work, showCheckboxes, selectedEntities
 				<a href={`/work/${work.bbid}`}>{name}</a>
 				{disambiguation}
 			</td>
+			<td>{work.author && <a href={`/author/${work.author.bbid}`}>{work.author.defaultAlias.name}</a>}</td>
```
Comment (MonkeyDo) [inline src/client/components/pages/entities/work-table.js:62]: As mentioned in the ORM PR, we will want an array or authors as works can have multiple "author wrote work" relationships.

**Comment by the-good-boy** [comment]
In the middleware ``loadWorkTableAuthors``, I have made use of a function called ``getRelationshipTargetBBIDByTypeId`` inspired from ``getRelationshipTargetByTypeId``. This function will give us a list of all workBBIDs of the edition. Then we will move to the ORM side.

Here, the [function](https://github.com/metabrainz/bookbrainz-data-js/pull/281/files#diff-dadc76e8af75fff4405eda2e1808fa591bbd045d1309d1206d198791317322b1) which we have used in the ORM will return us an array of objects containing authorAlias, authorBBID, and workBBID.

After this, I have grouped them by workBBID and stored them like a map in ``entity.authorsData``. 

After this, when we move to the Editions page, I have made use of a helper function called ``addAuthorsDataToWorks``, which will add the authorsData to the individual work objects, which will then be accessed in the Work-Table component.

**Review by MonkeyDo — COMMENTED** [review]
Looking really nice !

File: src/server/helpers/middleware.ts [hunk 2]
```diff
@@ -104,6 +105,16 @@ export function loadSeriesItems(req: $Request, res: $Response, next: NextFunctio
 	}
 }
 
+export async function loadWorkTableAuthors(req: $Request, res: $Response, next: NextFunction) {
+	const {orm}: any = req.app.locals;
+	const {entity} = res.locals;
+	const workBBIDs = getRelationshipTargetBBIDByTypeId(entity, 10);
+	const authorsData = await orm.func.work.loadAuthorNames(orm, workBBIDs);
```
[the remark this answers is not in the document]
Comment (the-good-boy) [reply src/server/helpers/middleware.ts]: Thanks for pointing that out! Made the changes :)

**Comment by MonkeyDo** [comment]
I've published a new version of the bookbrainz-data package for you to install in this PR: v2.13.0
https://github.com/metabrainz/bookbrainz-data-js/releases/tag/v2.13.0

**Comment by MonkeyDo** [comment]
I took the liberty prior to merging to squash some commits and fix merge conflicts, all related to updating bookbrainz-data version and the yarn lockfile.

**Comment by MonkeyDo** [comment]
CI tests are failing; apparently one of the tests is throwing because in our new SQL query there are no work bbids:
https://github.com/metabrainz/bookbrainz-site/runs/5395848930?check_suite_focus=true#step:10:935

> and work.bbid in () - syntax error at or near ")"

I'm not sure how that happens, considering we should be returning early if the array of bbids is empty. 
I can help investigate tomorrow if you'd like.

**Comment by the-good-boy** [comment]
@MonkeyDo I'm travelling rn(going back to college after two years). Will definitely look into this, as soon as I get settled.

**Comment by MonkeyDo** [comment]
I was just too puzzled to let it go, had to figure out how that was possible 😄 
After an hour of banging my head, it turns out I messed something up when I deployed the package and version 2.13.0 has the code for loadAuthorNames at the point of [this commit](https://github.com/metabrainz/bookbrainz-data-js/pull/281/commits/da50471ff05f5135ae2d5acd369d9f26206423f5), before we added the check for empty workBBIDs array.

I just redeployed the package and with the new version (2.13.1) tests are passing locally.
I'll push a commit and wait until tests pass, then we should be good to go.

**Comment by MonkeyDo** [comment]
Thanks @the-good-boy for the great feature, and for sticking with the long review process !

**Pull request merged**